// even-toolkit/dist/glasses/glass-mode.js
/**
 * Mode encoding for G2 glasses navigation.
 *
 * Packs a mode identifier + offset into a single highlightedIndex number.
 * E.g. { buttons: 0, read: 100, links: 200 } means:
 *   - index 0-99: buttons mode, offset = index
 *   - index 100-199: read mode, offset = index - 100
 *   - index 200+: links mode, offset = index - 200
 */
/**
 * Create a mode encoder from a mapping of mode names to base values.
 * Modes are matched by checking the index against bases in descending order.
 *
 * @param modes Record mapping mode names to their base values.
 *              All bases must be unique non-negative integers.
 *              E.g. { buttons: 0, scroll: 100, steps: 200 }
 */
export function createModeEncoder(modes) {
    // Sort entries by base value descending for decode matching
    const entries = Object.entries(modes)
        .sort((a, b) => b[1] - a[1]);
    function decode(index) {
        for (const [mode, base] of entries) {
            if (index >= base) {
                return { mode, offset: index - base };
            }
        }
        // Fallback to the lowest base mode
        const last = entries[entries.length - 1];
        return { mode: last[0], offset: index - last[1] };
    }
    return {
        encode(mode, offset = 0) {
            return modes[mode] + offset;
        },
        decode,
        getMode(index) {
            return decode(index).mode;
        },
        getOffset(index) {
            return decode(index).offset;
        },
        getBase(mode) {
            return modes[mode];
        },
    };
}
