// even-toolkit/dist/web/components/button.js
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from 'react';
import { cva } from 'class-variance-authority';
import { cn } from '../utils/cn';
const buttonVariants = cva('inline-flex items-center justify-center gap-4 rounded-[6px] font-normal transition-colors cursor-pointer disabled:opacity-50 disabled:pointer-events-none text-[17px] tracking-[-0.17px]', {
    variants: {
        variant: {
            highlight: 'bg-accent text-text-highlight hover:opacity-90',
            default: 'bg-surface text-text hover:bg-surface-light',
            ghost: 'text-text-dim hover:text-text hover:bg-surface-light',
            danger: 'bg-surface text-negative hover:bg-surface-light',
            secondary: 'bg-surface-light text-text-dim hover:text-text hover:bg-surface-lighter',
        },
        size: {
            sm: 'h-9 px-4 text-[15px] tracking-[-0.15px]',
            default: 'h-12 px-4',
            lg: 'h-12 px-6',
            icon: 'h-9 w-9 text-[15px]',
        },
    },
    defaultVariants: {
        variant: 'highlight',
        size: 'default',
    },
});
const Button = React.forwardRef(({ className, variant, size, type = 'button', ...props }, ref) => (_jsx("button", { ref: ref, type: type, className: cn(buttonVariants({ variant, size, className })), ...props })));
Button.displayName = 'Button';
export { Button, buttonVariants };
