// even-toolkit/dist/web/components/chat.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from 'react';
import { cn } from '../utils/cn';
function ChatThinking({ content, className }) {
    const [expanded, setExpanded] = React.useState(false);
    const summary = content.length > 80 ? content.slice(0, 80) + '...' : content;
    return (_jsxs("button", { type: "button", onClick: () => setExpanded(!expanded), className: cn('text-left w-full', className), children: [_jsxs("div", { className: "flex items-center gap-1 text-[13px] tracking-[-0.13px] text-text-dim cursor-pointer hover:text-text transition-colors", children: [_jsx("span", { children: expanded ? '\u25BE' : '\u25B8' }), _jsx("span", { className: "italic", children: "Thinking..." })] }), expanded && (_jsx("div", { className: "border-l-2 border-accent pl-3 py-1 mt-1 text-[13px] tracking-[-0.13px] text-text-dim italic whitespace-pre-wrap", children: content })), !expanded && (_jsx("div", { className: "border-l-2 border-accent pl-3 py-1 mt-1 text-[11px] tracking-[-0.11px] text-text-muted truncate", children: summary }))] }));
}
function ChatCodeBlock({ language, content, className }) {
    const lines = content.split('\n');
    return (_jsxs("div", { className: cn('bg-surface-lighter rounded-[6px] overflow-hidden', className), children: [language && (_jsx("div", { className: "px-3 py-1.5 text-[11px] tracking-[-0.11px] text-text-muted border-b border-border/30", children: language })), _jsx("div", { className: "overflow-x-auto", children: _jsx("pre", { className: "px-3 py-2 font-mono text-[13px] tracking-[-0.13px] text-text leading-5", children: lines.map((line, i) => (_jsxs("div", { className: "flex", children: [_jsx("span", { className: "w-8 shrink-0 text-right pr-3 text-text-muted select-none", children: i + 1 }), _jsx("span", { className: "whitespace-pre-wrap break-all", children: line })] }, i))) }) })] }));
}
function ChatDiff({ content, className }) {
    const lines = content.split('\n');
    return (_jsx("div", { className: cn('rounded-[6px] overflow-hidden font-mono text-[13px] tracking-[-0.13px]', className), children: lines.map((line, i) => {
            let bg = '';
            let textColor = 'text-text';
            if (line.startsWith('@@')) {
                bg = 'bg-accent/10';
                textColor = 'text-accent';
            }
            else if (line.startsWith('+')) {
                bg = 'bg-positive/10';
                textColor = 'text-positive';
            }
            else if (line.startsWith('-')) {
                bg = 'bg-negative/10';
                textColor = 'text-negative';
            }
            return (_jsx("div", { className: cn('px-3 py-0.5 whitespace-pre-wrap', bg, textColor), children: line }, i));
        }) }));
}
function ChatToolCall({ toolCall, className }) {
    const [expanded, setExpanded] = React.useState(false);
    return (_jsxs("div", { className: cn('bg-surface-lighter rounded-[6px] overflow-hidden', className), children: [_jsxs("button", { type: "button", onClick: () => setExpanded(!expanded), className: "w-full flex items-center gap-2 px-3 py-2 cursor-pointer hover:bg-surface-light/50 transition-colors", children: [toolCall.status === 'running' && (_jsx("div", { className: "w-3 h-3 animate-spin rounded-full border border-surface-lighter border-t-accent shrink-0" })), toolCall.status === 'complete' && (_jsxs("svg", { viewBox: "0 0 12 12", className: "w-3 h-3 text-positive shrink-0", children: [_jsx("rect", { x: 2, y: 6, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 4, y: 8, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 6, y: 6, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 8, y: 4, width: 2, height: 2, fill: "currentColor" })] })), toolCall.status === 'error' && (_jsxs("svg", { viewBox: "0 0 12 12", className: "w-3 h-3 text-negative shrink-0", children: [_jsx("rect", { x: 2, y: 2, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 4, y: 4, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 6, y: 6, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 8, y: 8, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 8, y: 2, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 6, y: 4, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 4, y: 6, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 2, y: 8, width: 2, height: 2, fill: "currentColor" })] })), !toolCall.status && (_jsxs("svg", { viewBox: "0 0 12 12", className: "w-3 h-3 text-text-dim shrink-0", children: [_jsx("rect", { x: 1, y: 3, width: 4, height: 2, fill: "currentColor" }), _jsx("rect", { x: 3, y: 5, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 5, y: 5, width: 4, height: 2, fill: "currentColor" }), _jsx("rect", { x: 7, y: 7, width: 2, height: 2, fill: "currentColor" })] })), _jsx("span", { className: "font-mono text-[13px] tracking-[-0.13px] text-text", children: toolCall.name }), _jsx("span", { className: "ml-auto text-[11px] text-text-muted", children: expanded ? '\u25BE' : '\u25B8' })] }), expanded && (_jsxs("div", { className: "border-t border-border/30 px-3 py-2 font-mono text-[13px] tracking-[-0.13px]", children: [toolCall.input && (_jsxs("div", { className: "mb-2", children: [_jsx("div", { className: "text-[11px] text-text-muted mb-1", children: "Input" }), _jsx("pre", { className: "whitespace-pre-wrap text-text-dim", children: toolCall.input })] })), toolCall.output && (_jsxs("div", { children: [_jsx("div", { className: "text-[11px] text-text-muted mb-1", children: "Output" }), _jsx("pre", { className: "whitespace-pre-wrap text-text-dim", children: toolCall.output })] }))] }))] }));
}
function ChatCommand({ command, output, className }) {
    return (_jsxs("div", { className: cn('font-mono text-[13px] tracking-[-0.13px]', className), children: [_jsxs("div", { className: "flex items-center gap-1 text-text", children: [_jsx("span", { className: "text-text-dim", children: "$" }), _jsx("span", { children: command })] }), output && (_jsx("pre", { className: "whitespace-pre-wrap text-text-dim mt-1", children: output }))] }));
}
function ChatError({ message, className }) {
    return (_jsx("div", { className: cn('bg-negative/10 text-negative rounded-[6px] px-3 py-2 text-[13px] tracking-[-0.13px]', className), children: message }));
}
// ─── Content Parser ─────────────────────────────────────────────
function parseAssistantContent(content) {
    const blocks = [];
    const lines = content.split('\n');
    let i = 0;
    let textBuffer = [];
    const flushText = () => {
        if (textBuffer.length > 0) {
            blocks.push(_jsx("span", { children: textBuffer.join('\n') }, `t-${blocks.length}`));
            textBuffer = [];
        }
    };
    while (i < lines.length) {
        const line = lines[i];
        // Code fence
        if (line.startsWith('```')) {
            flushText();
            const lang = line.slice(3).trim() || undefined;
            const codeLines = [];
            i++;
            while (i < lines.length && !lines[i].startsWith('```')) {
                codeLines.push(lines[i]);
                i++;
            }
            blocks.push(_jsx(ChatCodeBlock, { language: lang, content: codeLines.join('\n') }, `c-${blocks.length}`));
            i++; // skip closing ```
            continue;
        }
        // Diff block (consecutive +/- lines)
        if ((line.startsWith('@@') || line.startsWith('+') || line.startsWith('-')) && i + 1 < lines.length) {
            const diffStart = i;
            const diffLines = [];
            while (i < lines.length && (lines[i].startsWith('@@') || lines[i].startsWith('+') || lines[i].startsWith('-') || lines[i].startsWith(' '))) {
                diffLines.push(lines[i]);
                i++;
                // Stop if we've captured a meaningful diff (at least 2 lines starting with +/-)
                if (diffLines.length > 1 && i < lines.length && !lines[i].startsWith('@@') && !lines[i].startsWith('+') && !lines[i].startsWith('-') && !lines[i].startsWith(' '))
                    break;
            }
            if (diffLines.length >= 2 && diffLines.some(l => l.startsWith('+') || l.startsWith('-'))) {
                flushText();
                blocks.push(_jsx(ChatDiff, { content: diffLines.join('\n') }, `d-${blocks.length}`));
                continue;
            }
            // Not a real diff, treat as text
            i = diffStart;
        }
        // Tool call line: >> toolName
        if (line.startsWith('>> ')) {
            flushText();
            const name = line.slice(3).trim();
            blocks.push(_jsx(ChatToolCall, { toolCall: { name, status: 'complete' } }, `tc-${blocks.length}`));
            i++;
            continue;
        }
        // Error line: ! message
        if (line.startsWith('! ')) {
            flushText();
            blocks.push(_jsx(ChatError, { message: line.slice(2) }, `e-${blocks.length}`));
            i++;
            continue;
        }
        // Command line: $ command
        if (line.startsWith('$ ')) {
            flushText();
            blocks.push(_jsx(ChatCommand, { command: line.slice(2) }, `cmd-${blocks.length}`));
            i++;
            continue;
        }
        textBuffer.push(line);
        i++;
    }
    flushText();
    return blocks;
}
function ChatBubble({ message, className }) {
    const { role, content, thinking, isStreaming, toolCalls, codeBlocks, diff, error, command, timestamp } = message;
    if (role === 'system') {
        return (_jsx("div", { className: cn('flex justify-center py-2', className), children: _jsx("span", { className: "text-[13px] tracking-[-0.13px] text-text-dim", children: content }) }));
    }
    const isUser = role === 'user';
    return (_jsx("div", { className: cn('flex', isUser ? 'justify-end' : 'justify-start', 'mb-3', className), children: _jsxs("div", { className: "max-w-[85%] flex flex-col gap-1.5", children: [thinking && _jsx(ChatThinking, { content: thinking }), _jsxs("div", { className: cn('px-4 py-3 text-[15px] tracking-[-0.15px]', isUser
                        ? 'bg-accent text-text-highlight rounded-[6px] rounded-br-none'
                        : 'bg-surface text-text rounded-[6px] rounded-bl-none'), children: [isUser ? content : (_jsx("div", { className: "flex flex-col gap-2", children: parseAssistantContent(content) })), isStreaming && (_jsx("span", { className: "inline-block w-[2px] h-[15px] bg-current ml-0.5 animate-pulse align-text-bottom" }))] }), codeBlocks?.map((cb, i) => (_jsx(ChatCodeBlock, { language: cb.language, content: cb.content }, `cb-${i}`))), diff && _jsx(ChatDiff, { content: diff }), command && _jsx(ChatCommand, { command: command }), toolCalls?.map((tc, i) => (_jsx(ChatToolCall, { toolCall: tc }, `tc-${i}`))), error && _jsx(ChatError, { message: error }), timestamp && (_jsx("span", { className: cn('text-[11px] tracking-[-0.11px] text-text-muted', isUser ? 'text-right' : 'text-left'), children: new Date(timestamp).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' }) }))] }) }));
}
function ChatInput({ value, onChange, onSend, placeholder = 'Type a message...', disabled, className }) {
    const textareaRef = React.useRef(null);
    const handleKeyDown = (e) => {
        if (e.key === 'Enter' && !e.shiftKey) {
            e.preventDefault();
            if (value.trim())
                onSend();
        }
    };
    // Auto-grow textarea
    React.useEffect(() => {
        const ta = textareaRef.current;
        if (ta) {
            ta.style.height = '0';
            ta.style.height = Math.min(ta.scrollHeight, 120) + 'px';
        }
    }, [value]);
    return (_jsxs("div", { className: cn('flex items-end gap-2 p-3 bg-bg', className), children: [_jsx("textarea", { ref: textareaRef, value: value, onChange: (e) => onChange(e.target.value), onKeyDown: handleKeyDown, placeholder: placeholder, disabled: disabled, rows: 1, className: "flex-1 bg-input-bg text-text rounded-[6px] px-4 py-3 text-[15px] tracking-[-0.15px] outline-none placeholder:text-text-dim resize-none min-h-[44px] max-h-[120px]" }), _jsx("button", { type: "button", onClick: onSend, disabled: disabled || !value.trim(), className: "shrink-0 h-11 w-11 rounded-[6px] bg-accent text-text-highlight flex items-center justify-center cursor-pointer disabled:opacity-50 disabled:pointer-events-none transition-colors", children: _jsxs("svg", { viewBox: "0 0 24 24", className: "w-5 h-5", fill: "none", children: [_jsx("rect", { x: 10, y: 4, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 10, y: 6, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 10, y: 8, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 10, y: 10, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 10, y: 12, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 10, y: 14, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 10, y: 16, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 8, y: 6, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 12, y: 6, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 6, y: 8, width: 2, height: 2, fill: "currentColor" }), _jsx("rect", { x: 14, y: 8, width: 2, height: 2, fill: "currentColor" })] }) })] }));
}
function ChatContainer({ messages, input, className }) {
    const listRef = React.useRef(null);
    const [autoScroll, setAutoScroll] = React.useState(true);
    // Auto-scroll to bottom
    React.useEffect(() => {
        if (autoScroll && listRef.current) {
            listRef.current.scrollTop = listRef.current.scrollHeight;
        }
    }, [messages, autoScroll]);
    const handleScroll = () => {
        const el = listRef.current;
        if (!el)
            return;
        const atBottom = el.scrollHeight - el.scrollTop - el.clientHeight < 40;
        setAutoScroll(atBottom);
    };
    return (_jsxs("div", { className: cn('flex flex-col h-full', className), children: [_jsx("div", { ref: listRef, onScroll: handleScroll, className: "flex-1 overflow-y-auto px-3 py-4", children: messages.map((msg) => (_jsx(ChatBubble, { message: msg }, msg.id))) }), input] }));
}
export { ChatContainer, ChatBubble, ChatInput, ChatThinking, ChatCodeBlock, ChatDiff, ChatToolCall, ChatCommand, ChatError };
