// even-toolkit/dist/web/components/scroll-picker.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from 'react';
import { cn } from '../utils/cn';
/* -------------------------------------------------------------------------- */
/*  Constants                                                                 */
/* -------------------------------------------------------------------------- */
const ITEM_HEIGHT = 40;
const VISIBLE_ITEMS = 5;
const CONTAINER_HEIGHT = ITEM_HEIGHT * VISIBLE_ITEMS; // 200px
const PADDING_ITEMS = 2;
const PADDING_HEIGHT = ITEM_HEIGHT * PADDING_ITEMS; // 80px
const SCROLL_DEBOUNCE = 100;
function Column({ options, selectedValue, onChange }) {
    const containerRef = React.useRef(null);
    const isUserScrolling = React.useRef(false);
    const scrollTimer = React.useRef(null);
    const isMounted = React.useRef(true);
    const selectedIndex = React.useMemo(() => Math.max(0, options.findIndex((o) => o.value === selectedValue)), [options, selectedValue]);
    // Scroll to selected value on mount and when selectedValue changes externally
    React.useEffect(() => {
        const el = containerRef.current;
        if (!el)
            return;
        // Only programmatically scroll when not user-initiated
        if (isUserScrolling.current)
            return;
        const targetTop = selectedIndex * ITEM_HEIGHT;
        el.scrollTo({ top: targetTop, behavior: 'instant' });
    }, [selectedIndex]);
    React.useEffect(() => {
        isMounted.current = true;
        return () => {
            isMounted.current = false;
            if (scrollTimer.current)
                clearTimeout(scrollTimer.current);
        };
    }, []);
    const handleScroll = React.useCallback(() => {
        isUserScrolling.current = true;
        if (scrollTimer.current)
            clearTimeout(scrollTimer.current);
        scrollTimer.current = setTimeout(() => {
            if (!isMounted.current)
                return;
            const el = containerRef.current;
            if (!el)
                return;
            const scrollTop = el.scrollTop;
            const closestIndex = Math.round(scrollTop / ITEM_HEIGHT);
            const clampedIndex = Math.max(0, Math.min(closestIndex, options.length - 1));
            // Snap to the closest item
            el.scrollTo({ top: clampedIndex * ITEM_HEIGHT, behavior: 'smooth' });
            const newValue = options[clampedIndex]?.value;
            if (newValue && newValue !== selectedValue) {
                onChange(newValue);
            }
            // Reset flag after snap completes
            setTimeout(() => {
                if (isMounted.current)
                    isUserScrolling.current = false;
            }, 150);
        }, SCROLL_DEBOUNCE);
    }, [options, selectedValue, onChange]);
    const handleItemClick = React.useCallback((index) => {
        const el = containerRef.current;
        if (!el)
            return;
        isUserScrolling.current = true;
        el.scrollTo({ top: index * ITEM_HEIGHT, behavior: 'smooth' });
        const newValue = options[index]?.value;
        if (newValue)
            onChange(newValue);
        setTimeout(() => {
            if (isMounted.current)
                isUserScrolling.current = false;
        }, 300);
    }, [options, onChange]);
    return (_jsxs("div", { ref: containerRef, onScroll: handleScroll, className: "flex-1 overflow-y-auto overscroll-contain scrollbar-none", style: {
            height: CONTAINER_HEIGHT,
            scrollSnapType: 'y mandatory',
            WebkitOverflowScrolling: 'touch',
            scrollbarWidth: 'none',
            msOverflowStyle: 'none',
        }, children: [_jsx("div", { style: { height: PADDING_HEIGHT }, "aria-hidden": true }), options.map((option, i) => {
                const distance = Math.abs(i - selectedIndex);
                const isSelected = i === selectedIndex;
                // Opacity gradient: selected=1, +-1=0.4, +-2=0.2, beyond=0
                let opacity = 0;
                if (isSelected)
                    opacity = 1;
                else if (distance === 1)
                    opacity = 0.4;
                else if (distance === 2)
                    opacity = 0.2;
                return (_jsx("div", { onClick: () => handleItemClick(i), className: cn('flex items-center justify-center cursor-pointer select-none transition-all duration-150', isSelected
                        ? 'bg-accent text-text-highlight rounded-[6px] font-normal'
                        : 'text-text-dim'), style: {
                        height: ITEM_HEIGHT,
                        scrollSnapAlign: 'center',
                        opacity: isSelected ? 1 : opacity,
                    }, children: _jsx("span", { className: "text-[17px] tracking-[-0.17px]", children: option.label }) }, option.value));
            }), _jsx("div", { style: { height: PADDING_HEIGHT }, "aria-hidden": true })] }));
}
/* -------------------------------------------------------------------------- */
/*  ScrollPicker                                                              */
/* -------------------------------------------------------------------------- */
function ScrollPicker({ open, onClose, title, columns, value, onValueChange }) {
    // Close on Escape
    React.useEffect(() => {
        if (!open)
            return;
        const handler = (e) => {
            if (e.key === 'Escape')
                onClose();
        };
        document.addEventListener('keydown', handler);
        return () => document.removeEventListener('keydown', handler);
    }, [open, onClose]);
    // Prevent body scroll when open
    React.useEffect(() => {
        if (!open)
            return;
        const prev = document.body.style.overflow;
        document.body.style.overflow = 'hidden';
        return () => {
            document.body.style.overflow = prev;
        };
    }, [open]);
    const handleColumnChange = React.useCallback((columnIndex, newVal) => {
        const next = [...value];
        next[columnIndex] = newVal;
        onValueChange(next);
    }, [value, onValueChange]);
    if (!open)
        return null;
    return (_jsxs("div", { className: "fixed inset-0 z-50 flex items-end justify-center", children: [_jsx("div", { className: "absolute inset-0 bg-overlay", onClick: onClose }), _jsxs("div", { className: "relative z-10 w-full max-w-[393px] bg-bg rounded-t-[6px] shadow-[0px_-4px_12px_0px_rgba(0,0,0,0.12)] pt-3", children: [_jsx("div", { className: "px-4 py-3", children: _jsx("h2", { className: "text-[17px] tracking-[-0.17px] font-normal text-text text-center", children: title }) }), _jsx("div", { className: "relative flex items-center px-3", style: { height: CONTAINER_HEIGHT }, children: columns.map((col, colIndex) => {
                            // Resolve separator: check current column first, then previous column
                            const sep = colIndex > 0
                                ? (col.separator ?? columns[colIndex - 1]?.separator)
                                : undefined;
                            return (_jsxs(React.Fragment, { children: [sep && (_jsx("div", { className: "flex items-center justify-center shrink-0 px-1", style: { height: CONTAINER_HEIGHT }, children: _jsx("span", { className: "text-[17px] tracking-[-0.17px] font-normal text-text", children: sep }) })), _jsx(Column, { options: col.options, selectedValue: value[colIndex] ?? col.options[0]?.value ?? '', onChange: (v) => handleColumnChange(colIndex, v) })] }, colIndex));
                        }) }), _jsx("div", { className: "px-3 pt-3", children: _jsx("button", { type: "button", onClick: onClose, className: "w-full h-12 rounded-[6px] bg-surface text-text text-[17px] tracking-[-0.17px] font-normal cursor-pointer transition-colors hover:bg-surface-light", children: "Done" }) }), _jsx("div", { className: "flex items-center justify-center h-[42px]", children: _jsx("div", { className: "w-[139px] h-[5px] rounded-full bg-text" }) })] })] }));
}
/* -------------------------------------------------------------------------- */
/*  DatePicker                                                                */
/* -------------------------------------------------------------------------- */
const MONTHS = [
    'January', 'February', 'March', 'April', 'May', 'June',
    'July', 'August', 'September', 'October', 'November', 'December',
];
function DatePicker({ open, onClose, title = 'Select Date', value, onValueChange, minYear = 1900, maxYear = 2100, }) {
    const dayOptions = React.useMemo(() => Array.from({ length: 31 }, (_, i) => ({
        value: String(i + 1),
        label: String(i + 1),
    })), []);
    const monthOptions = React.useMemo(() => MONTHS.map((m, i) => ({
        value: String(i + 1),
        label: m,
    })), []);
    const yearOptions = React.useMemo(() => Array.from({ length: maxYear - minYear + 1 }, (_, i) => ({
        value: String(minYear + i),
        label: String(minYear + i),
    })), [minYear, maxYear]);
    const columns = React.useMemo(() => [
        { options: dayOptions },
        { options: monthOptions },
        { options: yearOptions },
    ], [dayOptions, monthOptions, yearOptions]);
    const pickerValue = React.useMemo(() => [String(value.day), String(value.month), String(value.year)], [value.day, value.month, value.year]);
    const handleChange = React.useCallback((v) => {
        onValueChange({
            day: parseInt(v[0], 10) || 1,
            month: parseInt(v[1], 10) || 1,
            year: parseInt(v[2], 10) || 2024,
        });
    }, [onValueChange]);
    return (_jsx(ScrollPicker, { open: open, onClose: onClose, title: title, columns: columns, value: pickerValue, onValueChange: handleChange }));
}
/* -------------------------------------------------------------------------- */
/*  TimePicker                                                                */
/* -------------------------------------------------------------------------- */
function TimePicker({ open, onClose, title = 'Select Time', value, onValueChange, format = '24h', }) {
    const is12h = format === '12h';
    const hourOptions = React.useMemo(() => {
        const max = is12h ? 12 : 23;
        const min = is12h ? 1 : 0;
        return Array.from({ length: max - min + 1 }, (_, i) => ({
            value: String(min + i),
            label: String(min + i).padStart(2, '0'),
        }));
    }, [is12h]);
    const minuteOptions = React.useMemo(() => Array.from({ length: 60 }, (_, i) => ({
        value: String(i),
        label: String(i).padStart(2, '0'),
    })), []);
    const periodOptions = React.useMemo(() => [
        { value: 'AM', label: 'AM' },
        { value: 'PM', label: 'PM' },
    ], []);
    const columns = React.useMemo(() => {
        const cols = [
            { options: hourOptions },
            { options: minuteOptions, separator: ':' },
        ];
        if (is12h) {
            cols.push({ options: periodOptions });
        }
        return cols;
    }, [hourOptions, minuteOptions, periodOptions, is12h]);
    const pickerValue = React.useMemo(() => {
        const vals = [String(value.hour), String(value.minute)];
        if (is12h)
            vals.push(value.period ?? 'AM');
        return vals;
    }, [value.hour, value.minute, value.period, is12h]);
    const handleChange = React.useCallback((v) => {
        const result = {
            hour: parseInt(v[0], 10) || 0,
            minute: parseInt(v[1], 10) || 0,
        };
        if (is12h) {
            result.period = v[2] ?? 'AM';
        }
        onValueChange(result);
    }, [is12h, onValueChange]);
    return (_jsx(ScrollPicker, { open: open, onClose: onClose, title: title, columns: columns, value: pickerValue, onValueChange: handleChange }));
}
/* -------------------------------------------------------------------------- */
/*  SelectionPicker                                                           */
/* -------------------------------------------------------------------------- */
function SelectionPicker({ open, onClose, title = 'Select', options, value, onValueChange, }) {
    const columns = React.useMemo(() => [{ options }], [options]);
    const pickerValue = React.useMemo(() => [value], [value]);
    const handleChange = React.useCallback((v) => {
        onValueChange(v[0] ?? '');
    }, [onValueChange]);
    return (_jsx(ScrollPicker, { open: open, onClose: onClose, title: title, columns: columns, value: pickerValue, onValueChange: handleChange }));
}
/* -------------------------------------------------------------------------- */
/*  Exports                                                                   */
/* -------------------------------------------------------------------------- */
export { ScrollPicker, DatePicker, TimePicker, SelectionPicker };
