// even-toolkit/dist/web/index.js
// even-toolkit/web — Shared web UI components for Even Realities apps
// Utils
export { cn } from './utils/cn';
// Icons
export { Icon, registerIcon, registerIcons, getIconNames } from './icons/index';
export { allIcons } from './icons/svg-icons';
export { svgCatalog, svgIconNames, svgIconCount } from './icons/svg-catalog';
// Primitives
export { Button, buttonVariants } from './components/button';
export { Card, cardVariants } from './components/card';
export { Badge, badgeVariants } from './components/badge';
export { Input } from './components/input';
export { Textarea } from './components/textarea';
export { Select } from './components/select';
export { Progress } from './components/progress';
export { StatusDot } from './components/status-dot';
export { Pill } from './components/pill';
export { Toggle } from './components/toggle';
export { MultiSelect } from './components/multi-select';
export { SegmentedControl } from './components/segmented-control';
export { Table, TableHeader, TableBody, TableRow, TableHead, TableCell } from './components/table';
export { Kbd } from './components/kbd';
export { Divider, dividerVariants } from './components/divider';
// Layout & Navigation
export { NavBar } from './components/nav-bar';
export { Page } from './components/page';
export { ScreenHeader } from './components/screen-header';
export { SectionHeader } from './components/section-header';
export { SettingsGroup } from './components/settings-group';
export { CategoryFilter } from './components/category-filter';
export { ListItem } from './components/list-item';
export { SearchBar } from './components/search-bar';
// Feedback & Overlay
export { TimerRing } from './components/timer-ring';
export { Dialog } from './components/dialog';
export { Toast, toastVariants } from './components/toast';
export { EmptyState } from './components/empty-state';
export { Loading } from './components/loading';
// New components from Figma Design Guidelines
export { NavHeader } from './components/nav-header';
export { BottomSheet } from './components/bottom-sheet';
export { CTAGroup } from './components/cta-group';
export { Tag, TagCarousel, TagCard } from './components/tag';
export { SliderIndicator, PageIndicator } from './components/indicator';
// Phase 1A: Foundational components
export { Checkbox } from './components/checkbox';
export { RadioGroup } from './components/radio-group';
export { Slider } from './components/slider';
export { Skeleton, skeletonVariants } from './components/skeleton';
export { InputGroup } from './components/input-group';
export { StepIndicator } from './components/step-indicator';
export { ConfirmDialog } from './components/confirm-dialog';
// Phase 1B: Premium components
export { ChatContainer, ChatBubble, ChatInput, ChatThinking, ChatCodeBlock, ChatDiff, ChatToolCall, ChatCommand, ChatError } from './components/chat';
export { Calendar } from './components/calendar';
export { Sparkline, LineChart, BarChart, PieChart, StatCard } from './components/chart';
// Phase 1C: Data visualization
export { Timeline } from './components/timeline';
export { StatGrid } from './components/stat-grid';
export { StatusProgress } from './components/status-progress';
// Phase 1D: Input & Media
export { FileUpload } from './components/file-upload';
export { VoiceInput, WaveformVisualizer } from './components/voice-input';
export { ImageGrid, ImageViewer } from './components/image-viewer';
export { AudioPlayer } from './components/audio-player';
// App Shell
export { AppShell } from './components/app-shell';
export { SideDrawer, DrawerTrigger, DrawerHeaderContext, useDrawerHeader } from './components/side-drawer';
export { DrawerShell } from './components/drawer-shell';
// Scroll Picker
export { ScrollPicker, DatePicker, TimePicker, SelectionPicker } from './components/scroll-picker';
