// even-toolkit/dist/stt/audio/buffer.js
import { float32ToWav } from './pcm-utils';
/** Audio accumulator — collects Float32 chunks and exports as WAV */
export function createAudioBuffer(config) {
    const sampleRate = config?.sampleRate ?? 16000;
    const maxSamples = (config?.maxSeconds ?? 30) * sampleRate;
    const chunks = [];
    let totalSamples = 0;
    function append(chunk) {
        if (totalSamples + chunk.length > maxSamples)
            return;
        chunks.push(chunk);
        totalSamples += chunk.length;
    }
    function getAll() {
        const out = new Float32Array(totalSamples);
        let offset = 0;
        for (const chunk of chunks) {
            out.set(chunk, offset);
            offset += chunk.length;
        }
        return out;
    }
    function getWav() {
        return float32ToWav(getAll(), sampleRate);
    }
    function clear() {
        chunks.length = 0;
        totalSamples = 0;
    }
    function duration() {
        return totalSamples / sampleRate;
    }
    return { append, getAll, getWav, clear, duration };
}
