// even-toolkit/dist/stt/providers/deepgram.js
import { sttLog } from '../debug';
export class DeepgramProvider {
    constructor() {
        this.type = 'deepgram';
        this.supportedModes = ['streaming'];
        this._state = 'idle';
        this.apiKey = '';
        this.language = 'en';
        this.modelId = 'nova-2';
        this.ws = null;
        this.suppressSocketEvents = false;
        this.transcriptCbs = [];
        this.stateCbs = [];
        this.errorCbs = [];
    }
    get state() {
        return this._state;
    }
    async init(config) {
        this.apiKey = config.apiKey ?? '';
        this.language = (config.language ?? 'en').split('-')[0]; // Deepgram wants 'en' not 'en-US'
        this.modelId = config.modelId ?? 'nova-2';
        if (!this.apiKey) {
            const err = { code: 'not-allowed', message: 'Deepgram API key is required', provider: this.type };
            this.emitError(err);
            throw new Error(err.message);
        }
    }
    start() {
        if (this.ws) {
            this.closeSocket(true);
        }
        this.suppressSocketEvents = false;
        const params = new URLSearchParams({
            model: this.modelId,
            language: this.language,
            interim_results: 'true',
            punctuate: 'true',
            encoding: 'linear16',
            sample_rate: '16000',
        });
        const url = `wss://api.deepgram.com/v1/listen?${params.toString()}`;
        this.ws = new WebSocket(url, ['token', this.apiKey]);
        this.ws.binaryType = 'arraybuffer';
        sttLog('deepgram: connecting to', url.substring(0, 60) + '...');
        this.ws.onopen = () => {
            sttLog('deepgram: connected');
            this.setState('listening');
        };
        this.ws.onmessage = (event) => {
            try {
                const data = JSON.parse(event.data);
                const alt = data.channel?.alternatives?.[0];
                const text = alt?.transcript ?? '';
                if (!text)
                    return;
                sttLog('deepgram: got', data.is_final ? 'FINAL' : 'interim', `"${text}"`);
                const transcript = {
                    text,
                    isFinal: data.is_final ?? false,
                    confidence: alt?.confidence ?? 0,
                    timestamp: Date.now(),
                };
                this.emitTranscript(transcript);
            }
            catch {
                // Non-JSON message, ignore
            }
        };
        this.ws.onerror = (event) => {
            if (this.suppressSocketEvents)
                return;
            sttLog('deepgram: WebSocket error', event);
            const err = {
                code: 'network',
                message: 'Deepgram WebSocket error',
                provider: this.type,
            };
            this.emitError(err);
            this.setState('error');
        };
        this.ws.onclose = () => {
            this.ws = null;
            if (this.suppressSocketEvents) {
                this.suppressSocketEvents = false;
                this.setState('idle');
                return;
            }
            if (this._state === 'listening') {
                this.setState('idle');
            }
        };
    }
    /** Send audio data to the Deepgram stream. Converts Float32 to Int16 (linear16). */
    sendAudio(data) {
        if (!this.ws || this.ws.readyState !== WebSocket.OPEN)
            return;
        if (data instanceof Float32Array) {
            // Convert Float32 [-1, 1] to Int16 PCM (Deepgram expects linear16)
            const int16 = new Int16Array(data.length);
            for (let i = 0; i < data.length; i++) {
                const s = Math.max(-1, Math.min(1, data[i]));
                int16[i] = s < 0 ? s * 0x8000 : s * 0x7fff;
            }
            this.ws.send(int16.buffer);
        }
        else if (data instanceof Int16Array) {
            this.ws.send(data.buffer);
        }
        else {
            this.ws.send(data);
        }
    }
    stop() {
        if (this.ws && this.ws.readyState === WebSocket.OPEN) {
            // Send close message per Deepgram protocol
            this.ws.send(JSON.stringify({ type: 'CloseStream' }));
        }
        this.closeSocket(true);
    }
    abort() {
        this.closeSocket(true);
    }
    dispose() {
        this.closeSocket(true);
        this.transcriptCbs = [];
        this.stateCbs = [];
        this.errorCbs = [];
    }
    onTranscript(cb) {
        this.transcriptCbs.push(cb);
        return () => { this.transcriptCbs = this.transcriptCbs.filter((c) => c !== cb); };
    }
    onStateChange(cb) {
        this.stateCbs.push(cb);
        return () => { this.stateCbs = this.stateCbs.filter((c) => c !== cb); };
    }
    onError(cb) {
        this.errorCbs.push(cb);
        return () => { this.errorCbs = this.errorCbs.filter((c) => c !== cb); };
    }
    // ── Private ──
    closeSocket(silent = false) {
        const socket = this.ws;
        if (socket) {
            this.ws = null;
            if (silent) {
                this.suppressSocketEvents = true;
                socket.onopen = null;
                socket.onmessage = null;
                socket.onerror = null;
                socket.onclose = null;
            }
            try {
                socket.close();
            }
            catch { /* ignore */ }
        }
        this.setState('idle');
    }
    setState(s) {
        if (this._state === s)
            return;
        this._state = s;
        for (const cb of this.stateCbs)
            cb(s);
    }
    emitTranscript(t) {
        for (const cb of this.transcriptCbs)
            cb(t);
    }
    emitError(e) {
        for (const cb of this.errorCbs)
            cb(e);
    }
}
